templates/fine-tune-function-calling/README.md

```diff
@@ -0,0 +1,419 @@
+    
+    
+    
+    
+
+
+If you notice, each sample has two entries: system and chat. This dataset is already formatted in specific way (e.g. using USER, \<|endoftext|\> and other tokens). To enable fine-tuning on various open source models we need to convert each row to a more general format like the OpenAI chat format, which is the preferred format for fine-tuning instruction-tuned models on Anyscale ([dataset format guide](https://docs.endpoints.anyscale.com/endpoints/fine-tuning/dataset-prep)). Let's first bring this dataset into the conversation format and inspect how that looks like:
+
+
+```python
+# initial preprocessing to get to the conversation format
+openai_fmt_ds = ray_ds.map(initial_mapper)
+pprint_example(openai_fmt_ds.take(1)[0], keys=["messages", "tools"]) # inspect one example
+```
+
+    2024-05-17 01:08:30,192	INFO streaming_executor.py:112 -- Starting execution of Dataset. Full logs are in /tmp/ray/session_2024-05-17_01-03-46_632348_53956/logs/ray-data
+    2024-05-17 01:08:30,193	INFO streaming_executor.py:113 -- Execution plan of Dataset: InputDataBuffer[Input] -> TaskPoolMapOperator[Map(initial_mapper)] -> LimitOperator[limit=1]
+
+
+
+    - Map(initial_mapper) 1:   0%|          | 0/1 [00:00<?, ?it/s]
+
+
+
+    - limit=1 2:   0%|          | 0/1 [00:00<?, ?it/s]
+
+
+
+    Running 0:   0%|          | 0/1 [00:00<?, ?it/s]
+
+
+    [91msystem: [0mYou are a helpful assistant.
+    [92muser: [0mHi, I need help with calculating the tip for my bill. The total amount is $50 and I want to give a 20% tip.
+    [94massistant: [0m[TOOL_CALLS] [ {"name": "calculate_tip", "arguments": '{"total": 50, "tip_percentage": 20}'} ] [/TOOL_CALLS]
+    [93mtool: [0m[{"tip_amount": 10}]
+    [94massistant: [0mBased on the total bill amount of $50 and a tip percentage of 20%, the tip amount you should give is $10. 
+    [95mTool list: [0m[{"name": "calculate_tip", "description": "Calculate the tip amount based on bill total and tip percentage", "parameters": {"type": "object", "properties": {"total": {"type": "number", "description": "The total bill amount"}, "tip_percentage": {"type": "number", "description": "The percentage of tip to be given"}}, "required": ["total", "tip_percentage"]}}]
+    
+
+
+Notice how we made use of special indicators "\[TOOL_CALLS\]" and "\[/TOOL_CALLS\]" to denote a tool call (as supposed a regular JSON output from the model). We'll now further process this conversation format and make it compatible with Anyscale Endpoints. The role "tool" will be converted to the role "user" with a special indicator to highlight that this is a tool response. Further, the tool list will be included in the system prompt with special indicators. The following code block handles the necessary preprocessing.
+
+
+```python
+# complete preprocessing step
+processed_ds = preprocess(ray_ds) 
```

Comment:
Use better names here:

so you'll be doing `anyscale_ds = openai_to_anyscale(ray_ds)`
Let's call the first one `openai_ds = glaive_to_openai(ray_ds)`

So these will be the two public functions. Everything else inside of it should be private function indicated by `_` in front of the name. e.g. `_filter_bad_examples`.

Also why is filtering not applied at when creating the openai dataset? It should be like the following flow:

```
raw_glaive_ds -> openai fmt -> filtered_openai -> anyscale fmt
```